### PR TITLE
Adding history module and implementing inmemory block storage and mix…

### PR DESCRIPTION
**Pull request by godmodegalactus**
…ed storage

### CODE REVIEW

File: core/src/structures/produced_block.rs [hunk 1]
```diff
@@ -22,3 +29,122 @@ pub struct ProducedBlock {
...
+        let txs = txs
+            .into_iter()
+            .filter_map(|tx| {
+                let Some(UiTransactionStatusMeta {
+                    err,
+                    compute_units_consumed,
+                    ..
+                }) = tx.meta
+                else {
+                    log::info!("Tx with no meta");
```
Comment (grooviegermanikus) [inline core/src/structures/produced_block.rs:54]: pls comment why this is a problem

pls write sth like "ignoring tx"

File: core/src/traits/block_storage_interface.rs [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
 #[async_trait]
 pub trait BlockStorageInterface: Send + Sync {
-    async fn save(&self, slot: Slot, block: UiConfirmedBlock, commitment: CommitmentLevel);
-    async fn get(&self, slot: Slot) -> Option<UiConfirmedBlock>;
+    // will save a block
+    async fn save(&self, slot: Slot, block: ProducedBlock, commitment: CommitmentLevel);
+    // will get a block
+    async fn get(&self, slot: Slot, config: RpcBlockConfig) -> Option<ProducedBlock>;
+    // will get range of slots that are stored in the storage
+    async fn get_range(&self) -> Range<Slot>;
```
Comment (grooviegermanikus) [inline core/src/traits/block_storage_interface.rs:14]: nit: method name could explicitly state that it's not about blocks but slots only..e.g. "get_slot_range"

File: history/src/block_stores/inmemory_block_store.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        let min_slot = match block_storage.first_key_value() {
+            Some((slot, _)) => *slot,
+            None => 0,
+        };
+        if slot > min_slot {
+            block_storage.insert(slot, block);
+            if block_storage.len() > self.number_of_blocks_to_store {
+                block_storage.remove(&min_slot)
+            } else {
+                None
```
Comment (grooviegermanikus) [inline history/src/block_stores/inmemory_block_store.rs:35]: looks wrong! shouldn't the method return the stored block?

File: history/src/block_stores/inmemory_block_store.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+#[async_trait]
+impl BlockStorageInterface for InmemoryBlockStore {
+    async fn save(
+        &self,
+        slot: Slot,
+        block: ProducedBlock,
+        _commitment: solana_sdk::commitment_config::CommitmentLevel,
+    ) {
+        self.store(slot, block).await;
```
Comment (grooviegermanikus) [inline history/src/block_stores/inmemory_block_store.rs:51]: do you want to allow to overwrite exiting slots?

File: history/src/block_stores/multiple_strategy_block_store.rs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+
+pub struct MultipleStrategyBlockStorage {
+    inmemory_for_storage: InmemoryBlockStore, // for confirmed blocks
+    persistent_block_storage: BlockStorageImpl, // for persistent block storage
+    faithful_rpc_client: Option<Arc<RpcClient>>, // to fetch legacy blocks from faithful
```
Comment (grooviegermanikus) [inline history/src/block_stores/multiple_strategy_block_store.rs:29]: that should go somewhere else.

File: history/src/block_stores/multiple_strategy_block_store.rs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+#[async_trait]
+impl BlockStorageInterface for MultipleStrategyBlockStorage {
+    async fn save(
+        &self,
+        slot: solana_sdk::slot_history::Slot,
+        block: ProducedBlock,
+        commitment: solana_sdk::commitment_config::CommitmentLevel,
+    ) {
+        if commitment == CommitmentLevel::Confirmed {
```
Comment (grooviegermanikus) [inline history/src/block_stores/multiple_strategy_block_store.rs:60]: nit: use match

File: history/src/block_stores/multiple_strategy_block_store.rs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                    self.inmemory_for_storage
+                        .save(slot, block.clone(), commitment)
+                        .await;
+                }
+            }
+            self.persistent_block_storage
+                .save(slot, block, commitment)
+                .await;
+        }
+        if slot > self.last_confirmed_slot.load(Ordering::Relaxed) {
```
Comment (grooviegermanikus) [inline history/src/block_stores/multiple_strategy_block_store.rs:83]: could use compare_exchange

File: history/src/block_stores/multiple_strategy_block_store.rs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            let range = self.inmemory_for_storage.get_range().await;
+            if range.contains(&slot) {
+                let block = self.inmemory_for_storage.get(slot, config).await;
+                if block.is_some() {
+                    return block;
+                }
+            }
+            // TODO: Define what data is expected that is definetly not in persistant block storage like data after epoch - 1
+            // check persistant block
+            let persistant_block_range = self.persistent_block_storage.get_range().await;
```
Comment (grooviegermanikus) [inline history/src/block_stores/multiple_strategy_block_store.rs:102]: typo: persistant

**Review by grooviegermanikus — CHANGES_REQUESTED** [review]

File: history/src/block_stores/inmemory_block_store.rs (hunk 3 above)
Comment (godmodegalactus) [reply history/src/block_stores/inmemory_block_store.rs:35]: this function was suppose to return the deleted block.

File: history/src/block_stores/inmemory_block_store.rs (hunk 4 above)
Comment (godmodegalactus) [reply history/src/block_stores/inmemory_block_store.rs:51]: Yes.

File: history/src/block_stores/multiple_strategy_block_store.rs (hunk 5 above)
Comment (godmodegalactus) [reply history/src/block_stores/multiple_strategy_block_store.rs:29]: Nah we want to fall back to faithful if the link is provided..

File: history/src/block_stores/multiple_strategy_block_store.rs (hunk 7 above)
Comment (godmodegalactus) [reply history/src/block_stores/multiple_strategy_block_store.rs:94]: compare exchange requires anyways to fetch the value and the value will be exchanged only if it is the same as current value which may add additional complexity.

**Review by aniketfuryrocks — APPROVED** [review]

**Pull request merged**